core/src/main/java/com/linecorp/armeria/server/ServerMetrics.java

```diff
@@ -16,36 +16,219 @@
 
 package com.linecorp.armeria.server;
 
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.LongAdder;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.google.common.base.MoreObjects;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 
 import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.DomainSocketAddress;
 
 import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.Tag;
 import io.micrometer.core.instrument.binder.MeterBinder;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
 
 /**
  * A class that holds metrics related server.
  */
 @UnstableApi
 public final class ServerMetrics implements MeterBinder {
 
-    private final LongAdder pendingHttp1Requests = new LongAdder();
-    private final LongAdder pendingHttp2Requests = new LongAdder();
-    private final LongAdder activeHttp1WebSocketRequests = new LongAdder();
-    private final LongAdder activeHttp1Requests = new LongAdder();
-    private final LongAdder activeHttp2Requests = new LongAdder();
+    private static final Logger logger = LoggerFactory.getLogger(ServerMetrics.class);
+
+    private static boolean warnedInvalidSocketAddress;
+
+    private final Iterable<ServerPort> ports;
+    private final boolean hasEphemeralPort;
+
+    private final Map<Integer, LongAdder> pendingHttp1Requests;
+    private final Map<Integer, LongAdder> pendingHttp2Requests;
+    private final Map<Integer, LongAdder> activeHttp1WebSocketRequests;
+    private final Map<Integer, LongAdder> activeHttp1Requests;
+    private final Map<Integer, LongAdder> activeHttp2Requests;
+    private final Map<String, LongAdder> domainSocketPendingHttp1Requests;
+    private final Map<String, LongAdder> domainSocketPendingHttp2Requests;
+    private final Map<String, LongAdder> domainSocketActiveHttp1WebSocketRequests;
+    private final Map<String, LongAdder> domainSocketActiveHttp1Requests;
+    private final Map<String, LongAdder> domainSocketActiveHttp2Requests;
```

Comment:
Optional) The implementation that holds many `Map`s in this class looks complex. How about introducing `ServerMetricsGroup` to encapsulate the complexity?
My suggestion:
```java
public interface ServerMetrics extends MeterBinder {
}

class ServerPortMetrics implements ServerMetrics {
    ...
}

class ServerMetricsGroup implements ServerMetrics {

    private Map<Object, ServerPortMetrics> delegates;
    ...
}
```

Reply:
I don't fully understand your intention.
So the `ServerMetricsGroup` has all the delegates?

Reply:
Right. I imagined the following implementation:
```java
class ServerMetricsGroup implements ServerMetrics {

  private Map<Object, ServerPortMetrics> delegates;

  void increasePendingHttp1Requests(SocketAddress socketAddress) {
      Object key = getPathOrPort(socketAddress);
      delegates.get(key).increasePendingHttp1Requests();
  }

  public long pendingHttp1Requests() {
      // Sum all pendingHttp1Requests in delegates and return
  }
}
```

Reply:
That looks much simpler. 👍 
Let me use that approach.

Reply:
On my second thought, I've realized that I didn't want to use the `Object` as a key because
- `InetSocketAddress` is used mostly for applications
- So the integer port needs to box and unbox.
 
 Let me develop your idea and avoid those unboxing.